Change the argument type of HTMLSlotElement.assign() into (Element or Text)

tetsuharuohzeki opened the pull request:
#### 48034229ee50a9abebeb9837701560cb2ad448a4
<pre>
Change the argument type of HTMLSlotElement.assign() into (Element or Text)
<a href="https://bugs.webkit.org/show_bug.cgi?id=243815">https://bugs.webkit.org/show_bug.cgi?id=243815</a>

Reviewed by Ryosuke Niwa.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api.html:
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assign):
* Source/WebCore/html/HTMLSlotElement.h:
* Source/WebCore/html/HTMLSlotElement.idl:

Canonical link: <a href="https://commits.webkit.org/253396@main">https://commits.webkit.org/253396@main</a>
</pre>

rniwa commented on Source/WebCore/html/HTMLSlotElement.cpp:
```diff
@@ -165,18 +166,26 @@ Vector<Ref<Element>> HTMLSlotElement::assignedElements(const AssignedNodesOption
+    HashSet<RefPtr<Node>> seenNodes;
+    m_manuallyAssignedNodes = WTF::compactMap(nodes, [&seenNodes](ElementOrText& node) -> std::optional<WeakPtr<Node>> {
+        auto mapper = [&seenNodes]<typename T>(RefPtr<T>& node) -> std::optional<WeakPtr<Node>> {
+            // RefPtr<Node>& node = derived;
```
What is this comment about? Remove?

tetsuharuohzeki replied:
I removed it in https://github.com/WebKit/WebKit/pull/3215/commits/ef32b1f3be32c721015225b5c02d3bfba44dde08

rniwa commented on Source/WebCore/html/HTMLSlotElement.idl:
```diff
@@ -30,7 +30,7 @@
     sequence<Node> assignedNodes(optional AssignedNodesOptions options);
     sequence<Element> assignedElements(optional AssignedNodesOptions options);
-    [EnabledBySetting=ImperativeSlotAPIEnabled] undefined assign(Node... nodes); // FIXME: This should be (Element or Text) instead of Node.
+    [EnabledBySetting=ImperativeSlotAPIEnabled] undefined assign((Element or Text)... nodes);
```
Can we add a test for this? In theory, we should be throwing in cases like this:
```js
const slot = document.createElement('slot');
slot.assign(document.createComment('foo'));
```

tetsuharuohzeki replied:
I added the test case to wpt.
https://github.com/web-platform-tests/wpt/pull/35443

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api.html:
```diff
@@ -285,4 +285,19 @@
+  }
+}, 'throw TypeError if the passed values are neither Element nor Text');
+
+</script>
```
Missing newline?

tetsuharuohzeki replied:
Thank you. I addressed it.

rniwa reviewed (approved):
r=me

tetsuharuohzeki committed:
Change the argument type of HTMLSlotElement.assign() into (Element or Text)
https://bugs.webkit.org/show_bug.cgi?id=243815

Reviewed by Ryosuke Niwa.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api.html:
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assign):
* Source/WebCore/html/HTMLSlotElement.h:
* Source/WebCore/html/HTMLSlotElement.idl:

Canonical link: https://commits.webkit.org/253396@main
